### PR TITLE
PriorityQueue refactor

### PR DESCRIPTION
**Pull request by wthorp**
https://storjlabs.atlassian.net/browse/V3-510

Work in Progress.  I wrapped PriorityQueue in another class to hide heap.  Realized it made sense to hide the XOR logic as well.  Tests stopped passing, tried to clean things up.

### CODE REVIEW

File: pkg/kademlia/queue.go [hunk 1]
```diff
@@ -4,57 +4,108 @@
...
+	pq *priorityQueue
+}
+
+//NewXorQueue returns a priorityQueue with priority based on XOR from targetBytes
+func NewXorQueue(nodes []*pb.Node, target dht.NodeID) XorQueue {
+	targetBytes := new(big.Int).SetBytes(target.Bytes())
+	pq := make(priorityQueue, len(nodes))
+	for i, node := range nodes {
+		pq[i] = &item{value: node, index: i}
+		pq[i].priority = new(big.Int).Xor(targetBytes, new(big.Int).SetBytes([]byte(node.GetId())))
```
Comment (egonelbre) [inline pkg/kademlia/queue.go]: Let's not use big.Int for this. Create a `nodeID [32]byte` and use that throughout. Can be a separate PR.

File: pkg/kademlia/queue.go [hunk 2]
```diff
@@ -4,57 +4,108 @@
...
+//We've encapsulated our priorityQueue{} implementation (below) so that you don't
+//have to know how the Golang heap API works (nor our priority XOR logic)
+
+//XorQueue is a priority queue where the priority is key XOR distance
+type XorQueue struct {
+	pq *priorityQueue
+}
+
+//NewXorQueue returns a priorityQueue with priority based on XOR from targetBytes
+func NewXorQueue(nodes []*pb.Node, target dht.NodeID) XorQueue {
```
Comment (egonelbre) [inline pkg/kademlia/queue.go]: `NewXorQueue(limit int) XorQueue` and use methods to add values.

File: pkg/kademlia/queue.go [hunk 3]
```diff
@@ -4,57 +4,108 @@
...
 )
 
-// An Item is something we manage in a priority queue.
-type Item struct {
+//We've encapsulated our priorityQueue{} implementation (below) so that you don't
+//have to know how the Golang heap API works (nor our priority XOR logic)
+
+//XorQueue is a priority queue where the priority is key XOR distance
+type XorQueue struct {
+	pq *priorityQueue
```
Comment (egonelbre) [inline pkg/kademlia/queue.go]: This doesn't need to be a pointer. naming it `items` might be clearer.

File: pkg/kademlia/queue.go [hunk 4]
```diff
@@ -4,57 +4,108 @@
...
+func (x XorQueue) Insert(node *pb.Node, target dht.NodeID) {
+	targetBytes := new(big.Int).SetBytes(target.Bytes())
+	heap.Push(x.pq, &item{
+		value:    node,
+		priority: new(big.Int).Xor(targetBytes, new(big.Int).SetBytes([]byte(node.GetId()))),
+	})
+}
+
+//PopClosest removed the closest priority node from the queue
+func (x XorQueue) PopClosest() (*pb.Node, dht.NodeID) {
```
Comment (egonelbre) [inline pkg/kademlia/queue.go]: The `Closest` part should be self-evident from the queue. So `Pop` should be sufficient.

File: pkg/kademlia/queue.go [hunk 5]
```diff
@@ -4,57 +4,108 @@
...
+	for i, node := range nodes {
+		pq[i] = &item{value: node, index: i}
+		pq[i].priority = new(big.Int).Xor(targetBytes, new(big.Int).SetBytes([]byte(node.GetId())))
+	}
+	heap.Init(&pq)
+	return XorQueue{pq: &pq} 
+}
+
+//Insert adds Node onto the queue
+func (x XorQueue) Insert(node *pb.Node, target dht.NodeID) {
```
Comment (egonelbre) [inline pkg/kademlia/queue.go]: `Pop` usually pairs with `Push`. So name it `Push`.

File: pkg/kademlia/queue.go [hunk 6]
```diff
@@ -4,57 +4,108 @@
...
+	targetBytes := new(big.Int).SetBytes(target.Bytes())
+	heap.Push(x.pq, &item{
+		value:    node,
+		priority: new(big.Int).Xor(targetBytes, new(big.Int).SetBytes([]byte(node.GetId()))),
+	})
+}
```
Comment (egonelbre) [inline pkg/kademlia/queue.go]: Handle the queue limit internally rather than pushing that responsibility to the user.

File: pkg/kademlia/queue.go (hunk 1 above)
Comment (wthorp) [reply pkg/kademlia/queue.go]: We currently derive our big.Int from dht.NodeId, which isn't fixed length.  We could use a MD5 to get a [16]byte or SHA-256 to get a [32]byte.  Thoughts on using a hash vs changing dht.NodeId?

File: pkg/kademlia/queue.go (hunk 2 above)
Comment (wthorp) [reply pkg/kademlia/queue.go]: agreed

File: pkg/kademlia/queue.go (hunk 3 above)
Comment (wthorp) [reply pkg/kademlia/queue.go]: agreed

File: pkg/kademlia/queue.go (hunk 4 above)
Comment (wthorp) [reply pkg/kademlia/queue.go]: enh, I went back and forth on naming conventions, push and pop imply a stack in my world

File: pkg/kademlia/queue.go (hunk 5 above)
Comment (wthorp) [reply pkg/kademlia/queue.go]: enh, I went back and forth on naming conventions, push and pop imply a stack in my world

File: pkg/kademlia/queue.go (hunk 6 above)
Comment (wthorp) [reply pkg/kademlia/queue.go]: agreed

File: pkg/kademlia/queue.go (hunk 4 above)
Comment (wthorp) [reply pkg/kademlia/queue.go]: I kept the "pop" because many priority queues implement a "peek", I wanted it to be clear that this was not a peek

File: pkg/kademlia/queue.go (hunk 5 above)
Comment (egonelbre) [reply pkg/kademlia/queue.go]: `Enqueue` and `Dequeue` is also a possibility. Or `Insert`, `Remove`. It's more about keeping the pair similar rather than the specific pair. As for why Push, Pop... because `container/heap` use them. Not feeling too strongly about which pair to use.

File: pkg/kademlia/queue.go (hunk 1 above)
Comment (coyle) [reply pkg/kademlia/queue.go]: BigInt also provides a built in Xor method that [32]byte doesn't

Comment (coyle) [reply pkg/kademlia/queue.go]: I'm also not sure why [32]byte is preferable to BigInt

Comment (egonelbre) [reply pkg/kademlia/queue.go]: I think it would end up more readable than using `big.Int`. As an example that exact thing could be:

```
pq[i].priority = target.Xor(nodeID(node))
```

File: pkg/kademlia/queue.go [hunk 7]
```diff
@@ -4,57 +4,99 @@
...
-type Item struct {
+//XorQueue is a priority queue where the priority is key XOR distance
+type XorQueue struct {
+	maxLen int
+	items  items
+}
+
+//NewXorQueue returns a items with priority based on XOR from targetBytes
+func NewXorQueue(size int) *XorQueue {
+	return &XorQueue{items: make(items, 0, size), maxLen: size}
```
Comment (coyle) [inline pkg/kademlia/queue.go]: this is limiting the queue to a maximum size but we want to let it grow larger right?

File: pkg/kademlia/queue.go [hunk 8]
```diff
@@ -4,57 +4,99 @@
...
+			heap.Push(&x.items, item)
+		}
+		heap.Init(&x.items)
+	}
+}
+
+//Closest removed the closest priority node from the queue
+func (x *XorQueue) Closest() (*pb.Node, dht.NodeID) {
+	item := *(heap.Pop(&x.items).(*item))
+	return item.value, item.priority
```
Comment (coyle) [inline pkg/kademlia/queue.go]: Priority isn't the nodeID though but the XOR distance from the nodeID and the target

File: pkg/kademlia/queue_test.go [hunk 9]
```diff
@@ -4,60 +4,41 @@
...
-		}
-		heap.Init(&v.pq)
-
-		i = 0
-		for v.pq.Len() > 0 {
-			item := heap.Pop(&v.pq).(*Item)
-			assert.Equal(t, big.NewInt(int64(v.expected[i])), item.priority)
-			i++
-		}
+func TestXorQueue(t *testing.T) {
```
Comment (coyle) [inline pkg/kademlia/queue_test.go]: can we test going over the limit ?

File: pkg/kademlia/queue.go (hunk 7 above)
Comment (wthorp) [reply pkg/kademlia/queue.go]: I think we use append to grow items later, which should safely resize the array.  see example https://play.golang.org/p/IBCJSqJaSf0.  All this really does is preallocate memory.

Comment (coyle) [reply pkg/kademlia/queue.go]: perfect I thought this limited the size it could grow to. Get my languages confused sometimes

File: pkg/kademlia/queue.go (hunk 8 above)
Comment (wthorp) [reply pkg/kademlia/queue.go]: Ok, this should probably explicitly return a big.Int, rather than the dht.NodeID interface.  We already have calculated the Xor before the priority get stored, so this should mean just changing the type.

File: pkg/kademlia/queue_test.go (hunk 9 above)
Comment (wthorp) [reply pkg/kademlia/queue_test.go]: added test, also added return of nil when queue is empty and test

File: pkg/kademlia/queue.go (hunk 7 above)
Comment (wthorp) [reply pkg/kademlia/queue.go]: yeah, I had to double check!

File: pkg/kademlia/queue.go [hunk 10]
```diff
@@ -4,57 +4,102 @@
...
+	}
+	//resize down if we grew too big
+	if x.items.Len() > x.maxLen {
+		olditems := x.items
+		x.items = items{}
+		for i := 0; i < x.maxLen && len(olditems) > 0; i++ {
+			item := heap.Pop(&olditems)
+			heap.Push(&x.items, item)
+		}
+		heap.Init(&x.items)
```
Comment (egonelbre) [inline pkg/kademlia/queue.go]: Note there are neat data-structures for this https://en.wikipedia.org/wiki/Double-ended_priority_queue, but we can do them in a later PR when necessary.
